src/Common/src/System/Net/Http/TlsCertificateExtensions.cs

```diff
@@ -0,0 +1,133 @@
+        /// </summary>
+        private static bool IsClientCertificate(X509Certificate2 cert)
+        {
+            Debug.Assert(cert != null, "certificate cannot be null");
+
+            bool hasClientCertExt = false;
+            bool forDigitalSignature = false;
+            bool noExtendedRestrictions = true;
+            bool noUsageRestrictions = true;
+            foreach(X509Extension extension in cert.Extensions)
+            {
+                // check if the extension is an enhanced usage ext.
+                // But do this only if needed. No point going over it, if we already have established that our cert has the
+                // required extension.
+                if (!hasClientCertExt)
+                {
+                    X509EnhancedKeyUsageExtension enhancedUsageExt = extension as X509EnhancedKeyUsageExtension;
+                    if (enhancedUsageExt != null)
+                    {
+                        noExtendedRestrictions = false;
+                        foreach(Oid oid in enhancedUsageExt.EnhancedKeyUsages)
+                        {
+                            if (string.Equals(ClientCertificateOid, oid.Value))
+                            {
+                                hasClientCertExt = true;
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                /// Check if the extension is a key usage extension.
+                /// No point going over it if we have already established that out cert has digital signature
+                if (!forDigitalSignature)
+                {
+                    X509KeyUsageExtension usageExt = extension as X509KeyUsageExtension;
+                    if (usageExt != null)
+                    {
+                        noUsageRestrictions = false;
+                        forDigitalSignature = (usageExt.KeyUsages & X509KeyUsageFlags.DigitalSignature) != 0;
+                    }
+                }
+
+                if(hasClientCertExt && forDigitalSignature)
+                {
+                    return true;
+                }
+            }
+            Debug.Assert(!hasClientCertExt || !forDigitalSignature, "if cert can be used as digital signature && it has ClientCertificateOid extension, it is a client cert");
+
+            return noExtendedRestrictions && noUsageRestrictions;
+        }
+
+        /// <summary>
+        ///   Returns a new collection containing valid client certificates from the given X509Certificate2Collection
+        /// </summary>
+        internal static bool TryFindClientCertificate(this X509Certificate2Collection certificates,
+                                                      ISet<string> allowedIssuers,
+                                                      out X509Certificate2 clientCertificate,
+                                                      out X509Chain clientCertChain)
+        {
```

Comment:
I am returning both clientCertificate as well as the certificate chain. ```clientCertChain[0]``` should be the certificate so there should be no reason for us to return the certificate as well. But ```clientCertChain[0]``` does not have a private key. I do not know if this is a bug or expected behavior or if I am missing something. That is the reason am sending out the certificate as well.